File: core/cyp/agents/risk_officer.py
"""风控官：在确定性风控引擎之上做 LLM 软评审。

铁律：**只能收紧，不能放宽**。无权批准被硬护栏否决的提案；只能把 approved/downsized
进一步升级为 rejected，或抬高 risk_score。无 LLM 时直接透传引擎结论（llm_reviewed=False）。
"""

from __future__ import annotations

from pydantic import BaseModel, Field

from cyp.agents.base import AgentContext
from cyp.contracts import AnalystReport, RiskAssessment, TradeProposal


class _RiskReview(BaseModel):
    risk_score: float = Field(ge=0.0, le=1.0)
    escalate_reject: bool = False
    notes: str = ""


class RiskOfficer:
    id = "risk_officer"

    async def run(
        self,
        proposal: TradeProposal,
        assessment: RiskAssessment,
        reports: list[AnalystReport],
        ctx: AgentContext,
    ) -> RiskAssessment:
        if assessment.verdict == "rejected":
            return assessment  # 已被硬护栏否决，无需软评审
        if not ctx.llm.enabled:
            return assessment  # 无 LLM：透传，llm_reviewed 保持 False

        drivers = "; ".join(f"{r.agent}:{r.stance}({r.confidence:.2f})" for r in reports)
        review = await ctx.llm.json(
            system=("你是加密交易风控官。只能收紧不能放宽：若发现 thesis 不自洽、极端行情/事件窗口、"
                    "或与已有敞口叠加同向风险，可 escalate_reject=true。给出 0-1 风险分与简短中文说明。"),
            user=f"提案：{proposal.side} {proposal.symbol} 仓位={proposal.size_quote} "
                 f"止损={proposal.stop_loss} 置信={proposal.confidence:.2f}\n分析：{drivers}",
            schema=_RiskReview,
        )
        if review is None:
            return assessment

        verdict = assessment.verdict
        violations = list(assessment.hard_violations)
        if review.escalate_reject:
            verdict = "rejected"
            violations.append(f"risk_officer: {review.notes or '软评审否决'}")
        return assessment.model_copy(update={
            "verdict": verdict,
            "hard_violations": violations,
            "risk_score": max(assessment.risk_score, review.risk_score),  # 取更保守
            "llm_notes": review.notes,
            "llm_reviewed": True,
        })

File: internal/agents/risk_officer.go
package agents

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"strings"

	"github.com/wangchaozhi/cyp-agent/internal/contracts"
)

var riskReviewSchema = json.RawMessage(`{
  "type":"object",
  "additionalProperties":false,
  "properties":{
    "risk_score":{"type":"number","minimum":0,"maximum":1},
    "escalate_reject":{"type":"boolean"},
    "notes":{"type":"string"}
  },
  "required":["risk_score"]
}`)

type riskReview struct {
	RiskScore      *float64 `json:"risk_score"`
	EscalateReject bool     `json:"escalate_reject"`
	Notes          string   `json:"notes"`
}

type RiskOfficer struct{}

func (RiskOfficer) Run(
	ctx context.Context,
	proposal contracts.TradeProposal,
	assessment contracts.RiskAssessment,
	reports []contracts.AnalystReport,
	agentContext AgentContext,
) (contracts.RiskAssessment, error) {
	if err := contextError(ctx); err != nil {
		return assessment, err
	}
	// The soft reviewer is structurally incapable of reviving a hard rejection.
	if assessment.Verdict == contracts.VerdictRejected || !agentContext.LLMEnabled() {
		return assessment, nil
	}
	drivers := make([]string, 0, len(reports))
	for _, report := range reports {
		drivers = append(drivers, fmt.Sprintf("%s:%s(%.2f)", report.Agent, report.Stance, report.Confidence))
	}
	var review riskReview
	err := agentContext.LLM.JSON(
		ctx,
		"你是加密交易风控官。只能收紧不能放宽：若发现 thesis 不自洽、极端行情/事件窗口、或与已有敞口叠加同向风险，可 escalate_reject=true。置信度偏低本身不是否决理由，应体现为抬高 risk_score。给出 0-1 风险分与简短中文说明。",
		fmt.Sprintf("提案：%s %s 仓位=%s 止损=%v 置信=%.2f\n分析：%s\n历史复盘经验：%s",
			proposal.Side, proposal.Symbol, proposal.SizeQuote, proposal.StopLoss, proposal.Confidence,
			strings.Join(drivers, "; "), recentLessons(agentContext.Lessons, 5)),
		riskReviewSchema,
		&review,
		false,
	)
	if err != nil {
		if contextError(ctx) != nil {
			return assessment, ctx.Err()
		}
		// Invalid JSON, provider failure, breaker, and budget exhaustion all
		// degrade to the deterministic hard-risk result.
		return assessment, nil
	}
	if review.RiskScore == nil || math.IsNaN(*review.RiskScore) || math.IsInf(*review.RiskScore, 0) || *review.RiskScore < 0 || *review.RiskScore > 1 {
		return assessment, nil
	}
	result := assessment
	result.HardViolations = append(contracts.List[string]{}, assessment.HardViolations...)
	if review.EscalateReject {
		result.Verdict = contracts.VerdictRejected
		note := truncateRunes(redactSensitive(strings.TrimSpace(review.Notes)), 1000)
		if note == "" {
			note = "软评审否决"
		}
		result.HardViolations = append(result.HardViolations, "risk_officer: "+note)
	}
	result.RiskScore = math.Max(assessment.RiskScore, *review.RiskScore)
	result.LLMNotes = redactSensitive(strings.TrimSpace(review.Notes))
	result.LLMNotes = truncateRunes(result.LLMNotes, 4000)
	result.LLMReviewed = true
	return result, nil
}
